Add RFC template and README file (#1)
uyuni-project/uyuni-rfc: README.md

# Uyuni RFCs

Many changes, including bug fixes and documentation improvements, can be implemented and reviewed via the normal GitHub pull request workflow. Some changes though are "substantial" - those should be put through an appropriate design process and produce a consensus among the Uyuni core team.

> The "RFC" (request for comments) process is intended to provide a consistent and controlled path for new features to enter the project. This process will be adjusted over time as more features are implemented and the community settles on specific approaches to feature development.

Reviewer: ```suggestion
The "RFC" (request for comments) process is intended to provide a consistent and controlled path for such substantial changes to enter the project. This process will be adjusted over time as more features are implemented and the community settles on specific approaches to feature development.
```

Author: I would agree, considering the RFC history, but OTOH, how do we create incentives for the community to keep future documents updated?

## TL;DR

* Fork https://github.com/uyuni-project/uyuni-rfc
* Copy `00000-template.md` to `active/0000-my-feature.md`
  - 'my-feature' should be as descriptive as possible.
  - please don't assign an RFC number yet.
* Write the content, based on the [RFC template](0000-template.md).
* Open a Pull Request and discuss it.
* When the discussion is finished and the PR ready to be merged, lookup the next free number in master and rename the file.

> Once the PR is merged, [implementation can start](#first-principles).

Reviewer: ```suggestion
Once the PR is merged, implementation can start.
```

Author: Why remove this link?

Reviewer: It points nowhere AFAICS. Am I missing anything?

Author: It points to the "First principles" heading/anchor below. The link should work.

Reviewer: Strange, for some reason it didn't at least in the GitHub preview

> > The Uyuni core team will work towards reviewing the set of open RFC pull requests and provide feedback on a weekly basis. All community members are invited to discuss the open pull requests.

Reviewer: ```suggestion
The Uyuni core team will work towards reviewing the set of open RFC pull requests and provide feedback on a weekly basis. All community members are invited to discuss the open pull requests.
```

## When to follow this process

You should consider using this process if you intend to make "substantial" changes to Uyuni or its documentation. Some examples of what typically requires an RFC are:

   - A new feature that creates new API surface area.
   - Architectural changes or introduction of new conventions.
   - Any disruptive changes in the user experience.
   - The removal of features that were already shipped.

The RFC process encourages discussions about a proposed feature during its early stages, in order to incorporate important constraints into the design. It is a great opportunity to get more eyeballs on your proposal before it becomes a part of a released version of Uyuni. Quite often, even proposals that seem "obvious" can be significantly improved once a wider group of interested people have a chance to weigh in.

#### Changes that do **NOT** require an RFC

  - Bug fixes.
  - Rephrasing, reorganizing or refactoring.
  - Improvements in test coverage or documentation.
  - UI improvements on existing features.

## First principles

- A RFC is considered 'active' when it gets merged into the `uyuni-rfc` repository.

> - Once an RFC becomes active, the feature can be implemented and submitted as a pull request to the Uyuni repository.

Reviewer 2: I am not sure if 'active' is the right word here, maybe we should say 'accepted' or 'agreed upon' instead? This way it would still be valid even if the implementation is incrementally drifting away from the RFC over time. It was 'accepted' at the time of merge, while it will most likely not stay 'active' forever in case we don't keep it up to date with the code. Also in some cases it might happen that people want to submit a PR with a prototype at the same time together with the RFC, which would be okay I guess.

Author: Changed all references to 'active' to 'accepted'.

- This does not mean the feature will be merged - only that the core team has agreed to it in principle.
- The fact that a given RFC has been accepted implies nothing about its implementation priority.
- Modifications to active RFC's can be done in follow-up pull requests.

> - We should strive to write each RFC in a way that reflects the final design of the feature; However, if during implementation things change, the RFC document should be updated accordingly.

Reviewer: While this last principle sounds great in theory, the bulk of our RFCs haven't been kept up-to-date with implementation changes, and so far the plan has been to keep them in the repo to basically reflect history.

It would be substantial work to update them all now - so either we wait publishing them until they are all up-to-date (and keep the maintenance cost going forward) or we change this rule into something like:

```suggestion
- We should strive to write each RFC in a way that reflects the final design of the feature; however, during implementation or afterwards things can change. The RFC only documents design decisions at the time it was merged.
```

Author: I would agree, considering the RFC history, but OTOH, how do we create incentives for the community to keep future documents updated?

Reviewer: Question is do we even want that?

Author: IMO yes since the RFCs are the only way we have to document the architecture and decisions behind relevant features. If what is on the document no longer holds true, I believe we should at least have sort of an addendum that briefly explains what changed and why. What are the counter arguments?

Reviewer: We need to maintain one more source of truth, and post-RFC discussions are documented anyway in PRs. This has a cost we did not pay so far - either we start and give the good example or we can't ask the community to do that.

FMPOV this can only come later, if at all.

Author: @renner @mcalmer what is your opinion?

### Implementation

Writing an RFC and implementing the feature are two distinct steps. The author of the RFC is not obligated to implement it. Of course, the RFC author (like any other developer) is welcome to contribute with an implementation after the RFC has been reviewed and merged.

> > If you are interested in implementing an 'active' RFC, but unable determine if someone else is already working on it, please ask at uyuni-devel@opensuse.org.

Reviewer: ```suggestion
If you are interested in implementing an 'active' RFC, but unable determine if someone else is already working on it, please ask at uyuni-devel@opensuse.org.
```

Reviewer 2: There is a small word missing here, it should be "unable **to** determine".
